src/content_mapblock.cpp

```diff
@@ -73,33 +73,49 @@ MapblockMeshGenerator::MapblockMeshGenerator(MeshMakeData *input, MeshCollector
 	blockpos_nodes = data->m_blockpos * MAP_BLOCKSIZE;
 }
 
-void MapblockMeshGenerator::useTile(int index, bool disable_backface_culling)
+void MapblockMeshGenerator::useTile(int index, int set_flags, int reset_flags, bool special)
 {
-	getNodeTileN(n, p, index, data, tile);
+	if (special)
+		getSpecialTile(index, &tile, p == data->m_crack_pos_relative);
+	else
+		getNodeTileN(n, p, index, data, tile);
 	if (!data->m_smooth_lighting)
 		color = encode_light(light, f->light_source);
 	for (int layer = 0; layer < MAX_TILE_LAYERS; layer++) {
-		tile.layers[layer].material_flags |= MATERIAL_FLAG_CRACK_OVERLAY;
-		if (disable_backface_culling)
-			tile.layers[layer].material_flags &= ~MATERIAL_FLAG_BACKFACE_CULLING;
+		tile.layers[layer].material_flags |= set_flags;
+		tile.layers[layer].material_flags &= ~reset_flags;
 	}
 }
 
-void MapblockMeshGenerator::useDefaultTile(bool set_color)
+void MapblockMeshGenerator::getTile(const v3s16& direction, TileSpec *tile)
```

Comment:
you can remove the const ref for v3s16 as it's 48b then its cost less to copy object than transport 64b reference